File: ios/Sources/OptimoveSDKPlugin/OptimoveSDKPlugin.swift
import Foundation
import Capacitor

@objc(OptimoveSDKPlugin)
public class OptimoveSDKPlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "OptimoveSDKPlugin"
    public let jsName = "OptimoveSDK"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "setCredentials", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "setUserId", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "setUserEmail", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "registerUser", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getVisitorId", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "signOutUser", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "reportEvent", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "reportScreenVisit", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "pushRequestDeviceToken", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "pushUnregister", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "inAppUpdateConsent", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "inAppGetInboxItems", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "inAppGetInboxSummary", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "inAppMarkAsRead", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "inAppMarkAllInboxItemsAsRead", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "inAppPresentInboxMessage", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "inAppDeleteMessageFromInbox", returnType: CAPPluginReturnPromise),
    ]

    private let implementation = OptimoveSDKImpl()

    // MARK: - Lifecycle

    override public func load() {
        implementation.setEventCallback { [weak self] eventName, data in
            self?.notifyListeners(eventName, data: data)
        }

        // Read config: capacitor.config.json first, Info.plist fallback
        let config = getConfig()

        implementation.initialize(
            optimoveCredentials: configString("optimoveCredentials", config: config),
            optimoveMobileCredentials: configString("optimoveMobileCredentials", config: config),
            inAppConsentStrategy: configString("inAppConsentStrategy", config: config) ?? "in-app-disabled",
            enableDeferredDeepLinking: configBool("enableDeferredDeepLinking", config: config),
            ddlCname: configString("ddlCname", config: config),
            enableDelayedInitialization: configBool("delayedInitialization.enable", config: config),
            delayedRegion: configString("delayedInitialization.region", config: config),
            delayedEnableOptimove: configBool("delayedInitialization.featureSet.enableOptimove", config: config),
            delayedEnableOptimobile: configBool("delayedInitialization.featureSet.enableOptimobile", config: config)
        )

        implementation.deliverPendingEvents()
    }

    // MARK: - Config Resolution (capacitor.config.json -> Info.plist fallback)

    /// Reads a string value: tries capacitor.config.json first, then Info.plist
    private func configString(_ key: String, config: PluginConfig) -> String? {
        // 1. Try capacitor.config.json (plugins.OptimoveSDK section)
        if let value = config.getString(key), !value.isEmpty {
            return value
        }
        // 2. Fall back to Info.plist
        if let value = Bundle.main.object(forInfoDictionaryKey: key) as? String, !value.isEmpty {
            return value
        }
        return nil
    }

    /// Reads a bool value: tries capacitor.config.json first, then Info.plist.
    /// Uses sentinel pattern to detect key presence in Capacitor config:
    /// if getBoolean(key, true) == getBoolean(key, false), the key exists with that value.
    /// If they differ, the key is absent (each returns its own default).
    private func configBool(_ key: String, config: PluginConfig) -> Bool {
        // 1. Try capacitor.config.json (handles both JSON boolean and string "true"/"false")
        let withTrue = config.getBoolean(key, true)
        let withFalse = config.getBoolean(key, false)
        if withTrue == withFalse {
            return withTrue
        }
        // 2. Fall back to Info.plist (supports both Bool and String values)
        if let value = Bundle.main.object(forInfoDictionaryKey: key) {
            if let boolValue = value as? Bool {
                return boolValue
            }
            if let stringValue = value as? String {
                return stringValue.caseInsensitiveCompare("true") == .orderedSame
            }
        }
        return false
    }

    // MARK: - User Identification

    @objc func setCredentials(_ call: CAPPluginCall) {
        implementation.setCredentials(
            optimoveCredentials: call.getString("optimoveCredentials"),
            optimobileCredentials: call.getString("optimobileCredentials")
        )
        call.resolve()
    }

    @objc func setUserId(_ call: CAPPluginCall) {
        guard let userId = call.getString("userId") else {
            call.reject("userId is required")
            return
        }
        implementation.setUserId(userId)
        call.resolve()
    }

    @objc func setUserEmail(_ call: CAPPluginCall) {
        guard let email = call.getString("email") else {
            call.reject("email is required")
            return
        }
        implementation.setUserEmail(email)
        call.resolve()
    }

    @objc func registerUser(_ call: CAPPluginCall) {
        guard let userId = call.getString("userId") else {
            call.reject("userId is required")
            return
        }
        guard let email = call.getString("email") else {
            call.reject("email is required")
            return
        }
        implementation.registerUser(userId: userId, email: email)
        call.resolve()
    }

    @objc func getVisitorId(_ call: CAPPluginCall) {
        call.resolve(["visitorId": implementation.getVisitorId()])
    }

    @objc func signOutUser(_ call: CAPPluginCall) {
        implementation.signOutUser()
        call.resolve()
    }

    // MARK: - Analytics

    @objc func reportEvent(_ call: CAPPluginCall) {
        guard let eventName = call.getString("event") else {
            call.reject("event name is required")
            return
        }
        let params = call.getObject("params") ?? [:]
        implementation.reportEvent(name: eventName, parameters: params)
        call.resolve()
    }

    @objc func reportScreenVisit(_ call: CAPPluginCall) {
        guard let screenName = call.getString("screenName") else {
            call.reject("screenName is required")
            return
        }
        implementation.reportScreenVisit(
            screenTitle: screenName,
            screenCategory: call.getString("screenCategory")
        )
        call.resolve()
    }

    // MARK: - Push

    @objc func pushRequestDeviceToken(_ call: CAPPluginCall) {
        implementation.pushRequestDeviceToken()
        call.resolve()
    }

    @objc func pushUnregister(_ call: CAPPluginCall) {
        implementation.pushUnregister()
        call.resolve()
    }

    // MARK: - In-App Messaging

    @objc func inAppUpdateConsent(_ call: CAPPluginCall) {
        guard let consented = call.getBool("consented") else {
            call.reject("consented is required")
            return
        }
        implementation.inAppUpdateConsent(consented: consented)
        call.resolve()
    }

    @objc func inAppGetInboxItems(_ call: CAPPluginCall) {
        DispatchQueue.global(qos: .userInitiated).async {
            let items = self.implementation.inAppGetInboxItems()
            call.resolve(["items": items])
        }
    }

    @objc func inAppGetInboxSummary(_ call: CAPPluginCall) {
        implementation.inAppGetInboxSummary(
            completion: { totalCount, unreadCount in
                call.resolve([
                    "totalCount": totalCount,
                    "unreadCount": unreadCount
                ])
            },
            failure: {
                call.reject("Could not get inbox summary")
            }
        )
    }

    @objc func inAppMarkAsRead(_ call: CAPPluginCall) {
        guard let messageId = call.getInt("id") else {
            call.reject("id is required")
            return
        }
        DispatchQueue.global(qos: .userInitiated).async {
            let result = self.implementation.inAppMarkAsRead(messageId: messageId)
            if result {
                call.resolve()
            } else {
                call.reject("Message not found or failed to mark as read")
            }
        }
    }

    @objc func inAppMarkAllInboxItemsAsRead(_ call: CAPPluginCall) {
        DispatchQueue.global(qos: .userInitiated).async {
            let result = self.implementation.inAppMarkAllInboxItemsAsRead()
            if result {
                call.resolve()
            } else {
                call.reject("Failed to mark all messages as read")
            }
        }
    }

    @objc func inAppPresentInboxMessage(_ call: CAPPluginCall) {
        guard let messageId = call.getInt("id") else {
            call.reject("id is required")
            return
        }
        DispatchQueue.global(qos: .userInitiated).async {
            let result = self.implementation.inAppPresentInboxMessage(messageId: messageId)
            call.resolve(["result": result])
        }
    }

    @objc func inAppDeleteMessageFromInbox(_ call: CAPPluginCall) {
        guard let messageId = call.getInt("id") else {
            call.reject("id is required")
            return
        }
        DispatchQueue.global(qos: .userInitiated).async {
            let result = self.implementation.inAppDeleteMessageFromInbox(messageId: messageId)
            if result {
                call.resolve()
            } else {
                call.reject("Message not found or not available")
            }
        }
    }
}

File: android/src/main/java/com/capacitor/optimove/sdk/OptimoveSDKPlugin.java
package com.capacitor.optimove.sdk;

import android.content.Context;

import androidx.annotation.Nullable;

import com.optimove.android.optimobile.InAppDeepLinkHandlerInterface;
import com.optimove.android.optimobile.OptimoveInApp;
import com.optimove.android.optimobile.PushMessage;

import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;

import org.json.JSONObject;

@CapacitorPlugin(name = "OptimoveSDK")
public class OptimoveSDKPlugin extends Plugin {

    @Nullable static OptimoveSDKPlugin instance;
    @Nullable static PushMessage pendingPush;
    @Nullable static String pendingActionId;
    @Nullable static JSObject pendingDDL;

    private final OptimoveSDK implementation = new OptimoveSDK();

    @Override
    public void load() {
        instance = this;
        deliverPendingEvents();
    }

    private void deliverPendingEvents() {
        if (pendingPush != null) {
            sendEvent("pushOpened", PushReceiver.pushMessageToJSObject(pendingPush, pendingActionId));
            pendingPush = null;
            pendingActionId = null;
        }
        if (pendingDDL != null) {
            sendEvent("deepLink", pendingDDL);
            pendingDDL = null;
        }
    }

    static void sendEvent(String eventName, JSObject data) {
        if (instance != null) {
            instance.notifyListeners(eventName, data);
        }
    }

    // ─── User Identification ──────────────────────────────

    @PluginMethod
    public void setCredentials(PluginCall call) {
        try {
            implementation.setCredentials(
                    call.getString("optimoveCredentials"),
                    call.getString("optimobileCredentials")
            );
            call.resolve();
        } catch (Exception e) {
            call.reject(e.getMessage(), e);
        }
    }

    @PluginMethod
    public void setUserId(PluginCall call) {
        String userId = call.getString("userId");
        if (userId == null) { call.reject("userId is required"); return; }
        try {
            implementation.setUserId(userId);
            call.resolve();
        } catch (Exception e) {
            call.reject(e.getMessage(), e);
        }
    }

    @PluginMethod
    public void setUserEmail(PluginCall call) {
        String email = call.getString("email");
        if (email == null) { call.reject("email is required"); return; }
        try {
            implementation.setUserEmail(email);
            call.resolve();
        } catch (Exception e) {
            call.reject(e.getMessage(), e);
        }
    }

    @PluginMethod
    public void registerUser(PluginCall call) {
        String userId = call.getString("userId");
        String email = call.getString("email");
        if (userId == null) { call.reject("userId is required"); return; }
        if (email == null) { call.reject("email is required"); return; }
        try {
            implementation.registerUser(userId, email);
            call.resolve();
        } catch (Exception e) {
            call.reject(e.getMessage(), e);
        }
    }

    @PluginMethod
    public void getVisitorId(PluginCall call) {
        String visitorId = implementation.getVisitorId();
        if (visitorId != null) {
            JSObject ret = new JSObject();
            ret.put("visitorId", visitorId);
            call.resolve(ret);
        } else {
            call.reject("visitor id is null");
        }
    }

    @PluginMethod
    public void signOutUser(PluginCall call) {
        try {
            implementation.signOutUser();
            call.resolve();
        } catch (Exception e) {
            call.reject(e.getMessage(), e);
        }
    }

    // ─── Analytics ────────────────────────────────────────

    @PluginMethod
    public void reportEvent(PluginCall call) {
        String eventName = call.getString("event");
        if (eventName == null) { call.reject("event is required"); return; }
        try {
            implementation.reportEvent(eventName, call.getObject("params"));
            call.resolve();
        } catch (Exception e) {
            call.reject(e.getMessage(), e);
        }
    }

    @PluginMethod
    public void reportScreenVisit(PluginCall call) {
        String screenName = call.getString("screenName");
        if (screenName == null) { call.reject("screenName is required"); return; }
        try {
            implementation.reportScreenVisit(screenName, call.getString("screenCategory"));
            call.resolve();
        } catch (Exception e) {
            call.reject(e.getMessage(), e);
        }
    }

    // ─── Push ─────────────────────────────────────────────

    @PluginMethod
    public void pushRequestDeviceToken(PluginCall call) {
        try {
            implementation.pushRequestDeviceToken();
            call.resolve();
        } catch (Exception e) {
            call.reject(e.getMessage(), e);
        }
    }

    @PluginMethod
    public void pushUnregister(PluginCall call) {
        try {
            implementation.pushUnregister();
            call.resolve();
        } catch (Exception e) {
            call.reject(e.getMessage(), e);
        }
    }

    // ─── In-App Messaging ─────────────────────────────────

    @PluginMethod
    public void inAppUpdateConsent(PluginCall call) {
        Boolean consented = call.getBoolean("consented");
        if (consented == null) { call.reject("consented is required"); return; }
        try {
            implementation.inAppUpdateConsent(consented);
            call.resolve();
        } catch (Exception e) {
            call.reject(e.getMessage(), e);
        }
    }

    @PluginMethod
    public void inAppGetInboxItems(PluginCall call) {
        try {
            JSObject result = implementation.inAppGetInboxItems();
            call.resolve(result);
        } catch (Exception e) {
            call.reject(e.getMessage(), e);
        }
    }

    @PluginMethod
    public void inAppGetInboxSummary(PluginCall call) {
        implementation.inAppGetInboxSummary(new OptimoveSDK.InboxSummaryCallback() {
            @Override
            public void onSuccess(int totalCount, int unreadCount) {
                JSObject ret = new JSObject();
                ret.put("totalCount", totalCount);
                ret.put("unreadCount", unreadCount);
                call.resolve(ret);
            }

            @Override
            public void onFailure() {
                call.reject("Could not get inbox summary");
            }
        });
    }

    @PluginMethod
    public void inAppMarkAsRead(PluginCall call) {
        Integer messageId = call.getInt("id");
        if (messageId == null) { call.reject("id is required"); return; }
        boolean result = implementation.inAppMarkAsRead(messageId);
        if (result) { call.resolve(); } else { call.reject("Message not found or failed to mark as read"); }
    }

    @PluginMethod
    public void inAppMarkAllInboxItemsAsRead(PluginCall call) {
        boolean result = implementation.inAppMarkAllInboxItemsAsRead();
        if (result) { call.resolve(); } else { call.reject("Failed to mark all messages as read"); }
    }

    @PluginMethod
    public void inAppPresentInboxMessage(PluginCall call) {
        Integer messageId = call.getInt("id");
        if (messageId == null) { call.reject("id is required"); return; }
        int result = implementation.inAppPresentInboxMessage(messageId);
        JSObject ret = new JSObject();
        ret.put("result", result);
        call.resolve(ret);
    }

    @PluginMethod
    public void inAppDeleteMessageFromInbox(PluginCall call) {
        Integer messageId = call.getInt("id");
        if (messageId == null) { call.reject("id is required"); return; }
        boolean result = implementation.inAppDeleteMessageFromInbox(messageId);
        if (result) { call.resolve(); } else { call.reject("Message not found or not available"); }
    }

    // ─── Inner Classes for Init Provider ──────────────────

    static class InAppDeepLinkHandler implements InAppDeepLinkHandlerInterface {
        @Override
        public void handle(Context context, InAppDeepLinkHandlerInterface.InAppButtonPress buttonPress) {
            JSObject data = new JSObject();
            try {
                data.put("deepLinkData", buttonPress.getDeepLinkData());
                data.put("messageId", buttonPress.getMessageId());
                JSONObject messageData = buttonPress.getMessageData();
                data.put("messageData", messageData == null ? JSObject.NULL : messageData);
            } catch (Exception e) {
                // noop
            }
            sendEvent("inAppDeepLink", data);
        }
    }

    static class InboxUpdatedHandler implements OptimoveInApp.InAppInboxUpdatedHandler {
        @Override
        public void run() {
            sendEvent("inAppInboxUpdated", new JSObject());
        }
    }
}
